### PR TITLE
Tune tbxlogger add images

### PR DESCRIPTION
**Pull request by simonsays1980**
## Why are these changes needed?

This PR enables users to provide in the `result` dicitonaries also image arrays that can be presented on TensorBoard like in the following example:

tensorboard

Images can be provided either as singleton in form of an `np.ndarray` with dimensions `(3, H, W)` or in form of a `4-d` `np.ndarray` with dimensions `(N, 3, H, W)` (in this case images get concatenated horizontally).

## Related issue number

#21954

As this issue is a https://github.com/ray-project/ray/labels/P1 https://github.com/ray-project/ray/labels/rllib the corresponding solution involves storing the images to the `episode.media` as this attribute of the episode is not summarized or appended in the `metrics.collect_episodes()` function.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by simonsays1980** [comment]
@Yard1 @krfricke My PR fails because of a flakey test. How can I re-run? Should I just mrge the master when it updates?

**Comment by JulianLorenz** [comment]
I also tried to apply a similar modification to TBXLoggerCallback. 
I discovered the following issue with image handling: 

When the image in episode.media gets processed by the JSONLoggerCallback, significant delays (minutes!) are introduced by the JSON logger. This is because the JSON logger needs to rewrite the log file at each logging call (see #21416) . I worked around this by disabling other Logger Callbacks by setting `TUNE_DISABLE_AUTO_CALLBACK_LOGGERS` to `1`.

Also I think it needs to be decided how to handle the case when images are logged by multiple episodes at one time in `summarize_episodes`. In my opinion, all images should be kept as it is currently implemented. The user can then decide when to log an image across multiple episodes in his algorithm class.

My modification of the TBXLoggerCallback.log_trial_result() function looks like this. I decided to make it more strict with the numpy array, to not automatically interpret any 3D array as image:
```python
            ...
            elif (isinstance(value, list) and len(value) > 0) or (
                    isinstance(value, np.ndarray) and value.size > 0
            ):
                valid_result[full_attr] = value

                # Check for list of images:
                if all(isinstance(v, np.ndarray) and v.ndim == 3 and v.shape[0] in [1, 3] for v in value):
                    if len(value) == 1:
                        # only one image
                        self._trial_writer[trial].add_image(full_attr, value[0], global_step=step)
                    else:
                        # Multiple images - stack them as tensorboard requires
                        imgs = np.stack(value)
                        self._trial_writer[trial].add_images(full_attr, imgs, global_step=step)
                    continue

                # Check for list of videos:
                if all(isinstance(v, np.ndarray) and v.ndim == 5 and v.shape[2] in [1,3] for v in value):
                    video = np.concatenate(value, axis=1)
                    self._trial_writer[trial].add_video(
                        full_attr, video, global_step=step, fps=20)
                    continue

                # Cover either a single video or a single image
                if isinstance(value, np.ndarray) and value.size > 0:
                    # Video - Must have 5 dimensions in NTCHW format:
                    # C must be either 1 for grayscale of 3 for RGB
                    if value.ndim == 5 and value.shape[2] in [1, 3]:
                        self._trial_writer[trial].add_video(
                            full_attr, value, global_step=step, fps=20
                        )
                        continue

                    # Image - Must have 3 dimensions in CHW format.
                    # C must be either 1 for grayscale of 3 for RGB
                    if value.ndim == 3 and value.shape[0] in [1, 3]:
                        self._trial_writer[trial].add_image(
                            full_attr, value, global_step=step
                        )
                        continue

                try:
                ...
```

File: python/ray/tune/logger/tensorboardx.py [hunk 1]
```diff
@@ -72,6 +72,24 @@ def on_result(self, result: Dict):
...
+                # Must be image
+                if isinstance(value, np.ndarray) and value.ndim == 3:
+                    self._file_writer.add_image(
+                        full_attr,
+                        value,
+                        global_step=step,
+                    )
+                    continue
+
+                # Must be multi-image
```
Comment (sven1977) [inline python/ray/tune/logger/tensorboardx.py:84]: Dumb question: What if this is a single video (t, w, h, c)?

File: python/ray/tune/logger/tensorboardx.py [hunk 2]
```diff
@@ -214,6 +232,24 @@ def log_trial_result(self, iteration: int, trial: "Trial", result: Dict):
...
+                # Must be image
+                if isinstance(value, np.ndarray) and value.ndim == 3:
+                    self._trial_writer[trial].add_image(
+                        full_attr,
+                        value,
+                        global_step=step,
+                    )
+                    continue
+
+                # Must be multi-image
```
Comment (sven1977) [inline python/ray/tune/logger/tensorboardx.py:244]:
```suggestion
                # Must be a batch of images.
```

File: python/ray/tune/logger/tensorboardx.py [hunk 3]
```diff
@@ -214,6 +232,24 @@ def log_trial_result(self, iteration: int, trial: "Trial", result: Dict):
             ):
                 valid_result[full_attr] = value
 
+                # Must be image
```
Comment (sven1977) [inline python/ray/tune/logger/tensorboardx.py:235]:
```suggestion
                # Must be a single image.
```

File: python/ray/tune/logger/tensorboardx.py (hunk 1 above)
Comment (sven1977) [inline python/ray/tune/logger/tensorboardx.py:84]:
```suggestion
                # Must be a batch of images.
```

File: python/ray/tune/logger/tensorboardx.py [hunk 4]
```diff
@@ -72,6 +72,24 @@ def on_result(self, result: Dict):
...
```
Comment (sven1977) [inline python/ray/tune/logger/tensorboardx.py:75]:
```suggestion
                # Must be a single image.
```

File: python/ray/tune/logger/tensorboardx.py (hunk 1 above)
Comment (simonsays1980) [reply python/ray/tune/logger/tensorboardx.py]: Following the definition of [`add_video()`](https://tensorboardx.readthedocs.io/en/latest/tensorboard.html#tensorboardX.SummaryWriter.add_video) only 5-dimensional inputs are accepted for this function. 

Do we anywhere pass data in a lower dimensional array into this function?

Comment (simonsays1980) [reply python/ray/tune/logger/tensorboardx.py]: @sven1977 Do you see any cases where this setup of separating arrays by dimension could fall on our feet?

**Review by sven1977 — APPROVED** [review]
LGTM!

**Pull request merged**